README
NAME
    cpan-outdated-fresh - Indicate out-of-date modules by walking the
    metacpan releases backwards

VERSION
    version 0.001006

SYNOPSIS
      cpan-outdated-fresh [--args]

        --age  TIMESPEC  The maximum age for a release (default: 7d)
         -a    TIMESPEC

        --develop        Include development releases in output
        --devel
        --dev

        --no-develop     Exclude development releases from output (default)
        --no-devel
        --no-dev

        --authorized     Show only authorized releases in output (default)
        --authed

        --no-authorized  Show even unauthorized releases in output
        --no-authed

  TIMESPEC
      <int seconds>
      <int><multiplier>

   multipliers
      s = second
      m = minute
      h = hour
      d = day
      w = 7 days
      M = 31 days
      Y = 365 days

DESCRIPTION
    This application bears much resemblance to cpan-outdated, but with a few
    important differences.

    *   MetaCPAN Driven

    *   "--dev" supported

    *   Temporal difference comparison.

  Temporal difference comparison
    One of the big features of using the "MetaCPAN" API, is the ability to
    filter results into a time bracket.

    As such, this application can tell you only about updates that happened
    to things you have installed "IF" they've been recently updated.

    For instance, one problem I kept facing was this:

      $ cpan-outdated
      # Broken::Package
      $ cpanm Broken::Package
      # Broken::Package is broken!
      ... wait a few days
      $ cpan-outdated
      # Broken::Package
      $ cpanm Broken::Package
      # Broken::Package is broken!
      ... doh
      ... repeat.

    With this app, as soon as "Broken::Package" falls outside the "age"
    window, you stop getting told about it.

    DU;DC - Didn't Update, Don't Care.

    This also has a benefit that if you run it very frequently, you can get
    a speed boost by narrowing the update window.

      $ time
      # 12:00
      $ cpan-outdated
      # install ALL the things
      ... wait a while
      $ time
      #  1:00
      $ cpan-outdated-fresh -a 1h
      # One update

    Essentially rewarding you for frequent use by being fast for frequent
    use.

   Downside
    Unfortunately, this optimization has a trade-off, namely, it scales very
    poorly vs time, due to exponential growth.

    "cpan-outdated" uses "02packages", which has its limits, and its
    inefficiencies, but the inefficiencies are effectively constant due to
    old packages getting pruned from "02packages" as newer versions replace
    them.

    So:

      cpan-outdated-fresh -a 30d

    Will take substantially more time to execute than

      cpan-outdated

    We've tried to combat this by trimming some cycles, but they may still
    prove over-zealous.

    However, this approach is still going to give you "--dev" results that
    "cpan-outdated" can't deliver.

SEE ALSO
    *   "App::cpanoutdated::fresh"

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for App-cpanoutdated-fresh

0.001007 2017-03-09T06:32:36Z 520e6a8

0.001006 2017-03-09T06:03:03Z 2e29975
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot
 - Bugfix: Make compatible with ES5

 [Dependencies::Stats]
 - Dependencies changed since 0.001005, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +15 ↑4 -3 (recommends: +1, suggests: ↑2)
 - runtime: ↓1 -2 (recommends: +1)
 - test: +1 (recommends: ↑1 ↓1)

0.001005 2014-09-19T07:03:13Z
 [Bugfix]
 - Handle fields being undef in string concat.

0.001004 2014-09-15T10:46:44Z
 [Bugfix]
 - Handle module->version being undef

0.001003 2014-09-15T02:48:14Z
 [Bugfix]
 - Be less overzealous about avoiding IO and catch upgrades where not all dists have version changes.

 [Dependencies::Stats]
 - Dependencies changed since 0.001002, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - test: (recommends: ↑1)

0.001002 2014-08-30T10:32:13Z
 [Bugfix]
 - Fix parsing timestamps to work as described.

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - test: +3 ↑1

 [Tests]
 - 50% coverage improvement.

0.001001 2014-08-30T00:39:10Z
 [00 Trivial]
 - metadata

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - develop: +1 ↑2 (suggests: ↑2)

 [Metadata]
 - Fix author

0.001000 2014-08-13T01:49:59Z
 - First version.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Indicate out-of-date modules by walking the metacpan releases backwards",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "App-cpanoutdated-fresh",
  "EXE_FILES" => [
    "bin/cpan-outdated-fresh"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "App::cpanoutdated::fresh",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Getopt::Long" => 0,
    "Module::Metadata" => 0,
    "Moo" => 0,
    "MooX::Lsub" => 0,
    "Path::ScanINC" => 0,
    "Pod::Usage" => 0,
    "Search::Elasticsearch" => 0,
    "strict" => 0,
    "version" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "HTTP::Tiny" => 0,
    "Test::Differences" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.001007",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Getopt::Long" => 0,
  "HTTP::Tiny" => 0,
  "Module::Metadata" => 0,
  "Moo" => 0,
  "MooX::Lsub" => 0,
  "Path::ScanINC" => 0,
  "Pod::Usage" => 0,
  "Search::Elasticsearch" => 0,
  "Test::Differences" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.96",
  "strict" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw( metacpan TIMESPEC cpan ));
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');

#$bundle->remove_policy('CodeLayout::RequireUseUTF8');
#$bundle->remove_policy('ErrorHandling::RequireCarping');
$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



lib/App/cpanoutdated/fresh.pm
use 5.006;    # our
use strict;
use warnings;

package App::cpanoutdated::fresh;

our $VERSION = '0.001007';

# ABSTRACT: Indicate out-of-date modules by walking the metacpan releases backwards

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Carp qw( croak );
use Moo qw( has );
use MooX::Lsub qw( lsub );
use Getopt::Long;
use Search::Elasticsearch;
use Module::Metadata;
use Path::ScanINC;
use Pod::Usage qw( pod2usage );
use version;





has ua => ( is => 'ro', predicate => 'has_ua' );
lsub trace => sub { undef };
lsub es => sub {
  my ($self) = @_;
  my %args = (
    nodes            => 'api.metacpan.org',
    cxn_pool         => 'Static::NoPing',
    send_get_body_as => 'POST',

    #    trace_to         => 'Stderr',
  );
  if ( $self->has_ua ) {
    $args{handle} = $self->ua;
  }
  if ( $self->trace ) {
    $args{trace_to} = 'Stderr';
  }
  return Search::Elasticsearch->new(%args);
};
lsub _sort       => sub { 'desc' };
lsub scroll_size => sub { 1000 };
lsub age         => sub { '7d' };
lsub age_seconds => sub {
  my ($self) = @_;
  my $table = {
    'm' => (60),
    'h' => ( 60 * 60 ),
    's' => (1),
    'd' => ( 24 * 60 * 60 ),
    'w' => ( 7 * 24 * 60 * 60 ),
    'M' => ( 31 * 24 * 60 * 60 ),
    'Y' => ( 365 * 24 * 60 * 60 ),
  };
  return $self->age + 0 if $self->age =~ /\A\d+([.]\d+)?\z/msx;
  if ( my ( $time, $multiplier ) = $self->age =~ /\A(\d+)([[:alpha:]]+)\z/msx ) {
    if ( not exists $table->{$multiplier} ) {
      croak("Unknown time multiplier <$multiplier>");
    }
    return $time * $table->{$multiplier};
  }
  croak( 'Cant parse age <' . $self->age . '>' );
};
lsub min_timestamp => sub {
  my ($self) = @_;
  return time() - $self->age_seconds;
};
lsub developer    => sub { undef };
lsub all_versions => sub { undef };
lsub authorized   => sub { 1 };
lsub _inc_scanner => sub { Path::ScanINC->new() };

sub _es_version {
  my ( $self, $wanted_version ) = @_;
  local $@ = undef;
  return eval { $self->es->VERSION($wanted_version); 1 };    ## no critic (RequireCheckingReturnValueOfEval)
}

sub _mk_scroll {
  my ($self) = @_;

  my $body = {
    query => {
      range => {
        'stat.mtime' => {
          gte => $self->min_timestamp,
        },
      },
    },
  };
  if ( not $self->developer or $self->authorized ) {
    $body->{filter} ||= {};
    $body->{filter}->{term} ||= {};
  }
  if ( not $self->developer ) {
    $body->{filter}->{term}->{'maturity'} = 'released';
  }
  if ( $self->authorized ) {
    $body->{filter}->{term}->{'authorized'}        = 'true';
    $body->{filter}->{term}->{'module.authorized'} = 'true';
  }

  my $fields = [
    qw(
      name distribution path
      stat.mtime module author
      authorized date indexed
      directory maturity release
      status version
      ),
  ];

  my %scrollargs = (
    scroll => '5m',
    index  => 'v0',
    type   => 'module',
    size   => $self->scroll_size,
    body   => $body,
    ( $self->_es_version(5) ? 'stored_fields' : 'fields' ) => $fields,
  );
  if ( not $self->_sort ) {
    $scrollargs{'search_type'} = 'scan';
  }
  else {
    $body->{sort} = { 'stat.mtime' => $self->_sort };
  }
  ( $self->_es_version(5) )
    ? ( require Search::Elasticsearch::Client::5_0::Scroll )
    : ( require Search::Elasticsearch::Scroll );

  return $self->es->scroll_helper(%scrollargs);
}

sub _check_fresh {
  my ( $self, $data_hash, $module ) = @_;
  return unless $module->{indexed} and $module->{authorized} and $module->{version};

  my (@parts) = split /::/msx, $module->{name};
  $parts[-1] .= '.pm';

  my $file = $self->_inc_scanner->first_file(@parts);
  return unless $file;

  my $mm = Module::Metadata->new_from_file($file);
  return if not $mm;

  my $mm_version = $mm->version;

  my $v = version->parse( $module->{version} );

  if ( not defined $v and not defined $mm_version ) {
    return;
  }
  if ( defined $v and not defined $mm_version ) {

    # noop, upstream got defined vs local, == upgrade
  }
  elsif ( not defined $v and defined $mm_version ) {

    # uhh, have version locally but not upstream, DONT upgrade
    return;
  }
  elsif ( $mm_version >= $v ) {
    return;
  }

  return {
    name      => $module->{name},
    cpan      => ( $v ? $v->stringify : q[0] ),
    release   => $data_hash->{release},
    installed => ( $mm_version ? $mm_version->stringify : q[0] ),
    meta      => $data_hash,
  };

}

sub _get_next {
  my ( $self, $scroll ) = @_;
  if ( not exists $self->{stash_cache} ) {
    $self->{stash_cache} = {};
  }
  if ( not exists $self->{upgrade_cache} ) {
    $self->{upgrade_cache} = {};
  }

  my $stash_cache   = $self->{stash_cache};
  my $upgrade_cache = $self->{upgrade_cache};

  while ( my $scroll_result = $scroll->next ) {
    return unless $scroll_result;
    my $data_hash = $scroll_result->{'_source'} || $scroll_result->{'fields'};

    my $cache_key = $data_hash->{path};
    my $upgrade_key =
        ( $data_hash->{author}       || 'NOAUTHOR' ) . q[/]
      . ( $data_hash->{distribution} || 'NODISTRIBUTION' ) . q[/]
      . ( $data_hash->{version}      || 'NOVERSION' );
    if ( $self->all_versions ) {
      $cache_key = $data_hash->{release};
    }

    #  pp($data_hash);
    next if exists $stash_cache->{$cache_key};
    next if not $self->developer and 'developer' eq $data_hash->{maturity};

    next if $data_hash->{path} =~ /\Ax?t\//msx;
    next unless $data_hash->{path} =~ /[.]pm\z/msx;
    next unless $data_hash->{module};
    next unless @{ $data_hash->{module} };
    for my $module ( @{ $data_hash->{module} } ) {
      my $fresh_data = $self->_check_fresh( $data_hash, $module );
      next unless $fresh_data;
      next if $upgrade_cache->{$upgrade_key};
      $upgrade_cache->{$upgrade_key} = 1;
      $stash_cache->{$cache_key}     = 1;
      return $fresh_data;
    }
    $stash_cache->{$cache_key} = 1;
  }
  return;
}









sub new_from_command {
  my ( $class, $defaults ) = @_;
  Getopt::Long::Configure('bundling');
  $defaults ||= {};
  my ( $help, $man );
  Getopt::Long::GetOptions(
    'age|a=s' => sub {
      my ( undef, $value ) = @_;
      $defaults->{age} = $value;
    },
    'develop|devel|dev!' => sub {
      my ( undef, $value ) = @_;
      if ($value) {
        $defaults->{developer} = 1;
        return;
      }
      $defaults->{developer} = undef;
    },
    'authorized|authed!' => sub {
      my ( undef, $value ) = @_;
      if ($value) {
        $defaults->{authorized} = 1;
      }
      else {
        $defaults->{authorized} = undef;
      }
    },
    'help|h|?' => \$help,
    'man'      => \$man,
  ) or do { $help = 1 };
  if ( $help or $man ) {
    if ($help) {
      return pod2usage( { -exitval => 1, }, );
    }
    return pod2usage( { -exitval => 1, -verbose => 2, }, );
  }
  return $class->new( %{$defaults} );
}









sub run {
  my ($self) = @_;
  my $iterator = $self->_mk_scroll;
  while ( my $result = $self->_get_next($iterator) ) {
    printf "%s\@%s\n", $result->{name}, $result->{cpan};
  }
  return 0;
}









sub run_command {
  my ($class) = @_;
  return $class->new_from_command->run();
}

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

App::cpanoutdated::fresh - Indicate out-of-date modules by walking the metacpan releases backwards

=head1 VERSION

version 0.001007

=head1 METHODS

=head2 new_from_command

Create an instance of this class parsing options from C<@ARGV>

  my $instance = App::cpanoutdated::fresh->new_from_command;

=head2 run

Execute the main logic and printing found modules to C<STDOUT>

  $object->run;

=head2 run_command

Shorthand for

  $class->new_from_command->run();

=for Pod::Coverage ua has_ua

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
